api/sanitize_ascii.go

```diff
@@ -0,0 +1,196 @@
+	"regexp"
+	"strings"
+)
+
+var jsonTypeRE = regexp.MustCompile(`[/+]json($|;)`)
+
+// GitHub servers return non-printable characters as their unicode code point values.
+// The values of \u0000 to \u001F represent C0 ASCII control characters and
+// the values of \u0080 to \u009F represent C1 ASCII control characters. These control
+// characters will be interpreted by the terminal, this behaviour can be used maliciously
+// as an attack vector, especially the control character \u001B. This function wraps
+// JSON response bodies in a ReadCloser that transforms C0 and C1 control characters
+// to their caret and hex notations respectively so that the terminal will not interpret them.
+func AddASCIISanitizer(rt http.RoundTripper) http.RoundTripper {
+	return &funcTripper{roundTrip: func(req *http.Request) (*http.Response, error) {
+		res, err := rt.RoundTrip(req)
+		if err != nil || !jsonTypeRE.MatchString(res.Header.Get("Content-Type")) {
+			return res, err
+		}
+		res.Body = &sanitizeASCIIReadCloser{rc: res.Body}
+		return res, err
+	}}
+}
+
+// sanitizeASCIIReadCloser implements the ReadCloser interface.
+type sanitizeASCIIReadCloser struct {
+	rc           io.ReadCloser
+	addBackslash bool
+	previousWin  []byte
+}
+
+// Read uses a sliding window alogorithm to detect C0 and C1
+// ASCII control sequences as they are read and replaces them
+// with equivelent inert characters. Characters that are not part
+// of a control sequence not modified.
+func (s *sanitizeASCIIReadCloser) Read(out []byte) (int, error) {
+	var readErr error
+	var outIndex int
+	var bufIndex int
+	var win []byte
+	buf := make([]byte, len(out))
+
+	bufLen, readErr := s.rc.Read(buf)
+	if readErr != nil && !errors.Is(readErr, io.EOF) {
+		if bufLen > 0 {
+			// Do not sanitize if there was a read error that is not EOF.
+			bufLen = copy(out, buf)
+		}
+		return bufLen, readErr
+	}
+
+	if s.previousWin != nil {
+		buf = append(s.previousWin, buf...)
+		bufLen += len(s.previousWin)
+	}
+
+	for {
+		remaining := min(6, (bufLen - bufIndex))
+		win = buf[bufIndex : bufIndex+remaining]
+		if remaining < 6 {
+			break
+		}
+
+		if win[0] == 92 && win[1] == 117 && win[2] == 48 && win[3] == 48 {
+			repl, _ := mapControlCharacterToCaret(win)
+			if s.addBackslash {
+				repl = append([]byte{92}, repl...)
+			}
+			l := len(repl)
+			for j := 0; j < l; j++ {
+				out[outIndex] = repl[j]
+				outIndex++
+			}
+			bufIndex += 6
+			s.addBackslash = false
+			continue
+		}
+
+		if win[0] == 92 {
+			s.addBackslash = !s.addBackslash
+		} else {
+			s.addBackslash = false
+		}
+
+		out[outIndex] = buf[bufIndex]
+		outIndex++
+		bufIndex++
+	}
+
+	if readErr != nil && errors.Is(readErr, io.EOF) {
+		remaining := bufLen - bufIndex
+		for j := 0; j < remaining; j++ {
+			out[outIndex] = win[j]
+			outIndex++
+			bufIndex++
+		}
+	} else {
+		s.previousWin = win
+	}
+
+	return outIndex, readErr
+}
+
+// Close closes the wrapped ReadCloser.
+func (s *sanitizeASCIIReadCloser) Close() error {
```

Comment:
Nit: since your struct is embedding an `io.ReadCloser` interface, I don't think you need to explicitly declare this method. Your type should already automatically have a `Close()` method and forward it to the underlying ReadCloser.

Reply:
I gave the `ReadCloser` wrapped by `sanitizeASCIIReadCloser` a name in the struct so Go won't promote the `Close` method to be on `sanitizeASCIIReadCloser`. Having said that I think it makes sense to making the `ReadCloser` actually embedded.